src/server/cluster/cluster_family.cc

```diff
@@ -748,6 +760,30 @@ void ClusterFamily::RemoveIncomingMigrations(const std::vector<MigrationInfo>& m
           return m.node_id == im->GetSourceID() && m.slot_ranges == im->GetSlots();
         });
     DCHECK(it != incoming_migrations_jobs_.end());
+    DCHECK(it->get() != nullptr);
+    IncomingSlotMigration& migration = *it->get();
+
+    // Flush non-owned migrations
+    SlotSet migration_slots(migration.GetSlots());
+    SlotSet removed = migration_slots.GetRemovedSlots(tl_cluster_config->GetOwnedSlots());
+
+    // First cancel socket, then flush slots, so that new entries won't arrive after we flush.
+    if (migration.GetState() != MigrationState::C_FINISHED) {
```

Comment:
state can be changed in the different thread so we don't know is it finished or not when you do cancel